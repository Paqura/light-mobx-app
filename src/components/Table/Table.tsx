import React, { useEffect, useState } from 'react';
import { useTableStore } from '../../stores';
import { observer } from 'mobx-react';
import { Table as RTable, Modal, Spinner } from 'react-bootstrap';
import { coronaService } from '../../services/Corona.service';
import { Chart } from '../Chart';

const formatCase = new Intl.NumberFormat().format;

const Table = () => {
  const [isLoading, setIsLoading] = useState(false);
  const [isModalShown, setIsModalShown] = useState(false);
  const [chartData, setChartData] = useState<number[]>([]);
  const { data, setSort, direction, setData } = useTableStore();

  const sortIcon = direction === 'up' ? '^' : direction === 'down' ? '_' : '*';

  useEffect(() => {
    const getData = async () => {
      setIsLoading(true);

      const data = await coronaService.getCountries();
      setData(data)

      setIsLoading(false);
    }

    getData();
  }, [setData]);

  const onChartOpen = (country: string) => async () => {
    const preparedCountry = country.toLowerCase();

    try {
      const { timeline } = await coronaService.getChartData(preparedCountry);
      const cases = Object.values(timeline.cases);
      setChartData(cases);

      setIsModalShown(true);
    } catch (error) {
      console.log(error.message || error);
    }
  };

  const closeModal = () => {
    setIsModalShown(false);
  }

  if (isLoading) {
    return <Spinner animation="border" />
  }

  return (
    <>
      <Modal onHide={closeModal} show={isModalShown} size="xl" animation={false}>
        <Chart data={chartData} />
      </Modal>
      <RTable striped bordered hover>
        <thead>
          <tr>
            <td>â„–</td>
            <td>Country</td>
            <td onClick={setSort}>
              Cases
              <span style={{ position: 'absolute' }}>{sortIcon}</span>
            </td>
            <td>Flag</td>
          </tr>
        </thead>
        <tbody>
          {data?.length > 0 && data.map((item, idx) => (
            <tr key={item.country}>
              <td>{idx}</td>
              <td onClick={onChartOpen(item.country)}>
                <button>{item.country}</button>
              </td>
              <td>{formatCase(item.cases)}</td>
              <td>
                <span>
                  <img src={item.countryInfo.flag} alt={`Flag of ${item.country}`} width="32px" />
                </span>
              </td>
            </tr>
          ))}
        </tbody>
      </RTable>
    </>
  )
}

const TableObserver = observer(Table);

export default TableObserver;