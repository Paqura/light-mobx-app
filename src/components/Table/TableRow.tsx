import React, { MouseEvent } from 'react'

const formatCase = new Intl.NumberFormat().format;

interface Props {
  idx: number;

  data: {
    country: string;
    cases: number;
    countryInfo: { flag: string };
  }

  onChartOpen: (country: string) => (evt: MouseEvent) => void;
}

const TableRow: React.FC<Props> = ({ data, idx, onChartOpen }) => {
  return (
    <tr>
      <td>{idx}</td>
      <td onClick={onChartOpen(data.country)}>
        <button>{data.country}</button>
      </td>
      <td>{formatCase(data.cases)}</td>
      <td>
        <span>
          <img src={data.countryInfo.flag} alt={`Flag of ${data.country}`} width="32px" />
        </span>
      </td>
    </tr>
  )
}

export default TableRow
