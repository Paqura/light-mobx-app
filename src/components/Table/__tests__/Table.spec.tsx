import React from 'react';
import renderer, { ReactTestRenderer } from 'react-test-renderer';
import Table from '../Table';

describe('Table', () => {
  let wrapper: ReactTestRenderer | null = null;

  beforeEach(() => {
    wrapper = renderer.create(
      <Table />
    );
  });

  it('should return snapshot', () => {
    expect(wrapper!.toJSON()).toMatchSnapshot();
  });

  afterEach(() => {
    wrapper = null;
  })
})