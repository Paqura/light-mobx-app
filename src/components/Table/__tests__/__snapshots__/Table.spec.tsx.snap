// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should return snapshot 1`] = `
<table
  className="table table-striped table-bordered"
>
  <thead>
    <tr>
      <td>
        â„–
      </td>
      <td>
        Country
      </td>
      <td
        onClick={[Function]}
      >
        Cases
        <span
          style={
            Object {
              "position": "absolute",
            }
          }
        >
          *
        </span>
      </td>
      <td>
        Deaths
      </td>
      <td>
        Flag
      </td>
    </tr>
  </thead>
  <tbody />
</table>
`;
